### PR TITLE
Add Databricks UC Volume Object Store

### PR DESCRIPTION
**Pull request by panchalhp-db**
# What does this PR do?
- Add support for Databricks Unity Catalog Volumes to upload and download data to UC

# What issue(s) does this change relate to?
N/A

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [x] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

# Testing
- Added unit tests
```
 python -m pytest tests/utils/object_store/test_uc_object_store.py
```
- Manually tested this with a sample script as below:
```
import composer.utils.file_helpers as fh
import tempfile
import os

def main():
    print('Testing object_store_test')
    uri = 'uc://Volumes/ml/automl/test-volume/'
    object_name = 'composer-file.txt'

    localfile1 = os.path.join(tempfile.gettempdir(), 'file1.txt')
    localfile2 = os.path.join(tempfile.gettempdir(), 'file2.txt')

    # cleanup
    if os.path.exists(localfile1):
        os.remove(localfile1)
    if os.path.exists(localfile2):
        os.remove(localfile2)

    # write something to file1
    with open(localfile1, 'w') as f:
        f.writelines('This is a sample composer file\n')

    object_store = fh.maybe_create_object_store_from_uri(uri=uri)

    # upload
    object_store.upload_object(object_name, localfile1)

    # download
    object_store.download_object(object_name, localfile2)

    # get size
    print(f'object_size={object_store.get_object_size(object_name)}')

    # validate that the contents of file1 and file2 are the same
    with open(localfile1, 'rb') as f:
        filecontent1 = f.readlines()
    with open(localfile2, 'rb') as f:
        filecontent2 = f.readlines()
    assert filecontent1 == filecontent2


if __name__ == '__main__':
    main()
```
Call the above script as below:
```
source venv/bin/activate; DATABRICKS_HOST=<host> DATABRICKS_TOKEN=<token> python manual-test.py
```
File is seen in UC Volume:
image
Verified contents of the file:
image

### Integration testing
```
DATABRICKS_HOST=<host> DATABRICKS_TOKEN=<host> python -m pytest -v -m remote --log-cli-level 10 tests/utils/object_store/test_uc_object_store.py::test_uc_object_store_integration
```
image

### CODE REVIEW

File: composer/utils/object_store/uc_object_store.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import uuid
+from typing import Callable
+
+from composer.utils.import_helpers import MissingConditionalImportError
+from composer.utils.object_store.object_store import ObjectStore
+
+log = logging.getLogger(__name__)
+
+
+class UCObjectStore(ObjectStore):
```
Comment (dakinggg) [inline composer/utils/object_store/uc_object_store.py:15]: can you add `list_objects` please?

File: composer/utils/file_helpers.py [hunk 2]
```diff
@@ -349,6 +349,8 @@ def maybe_create_object_store_from_uri(uri: str) -> Optional[ObjectStore]:
         return GCSObjectStore(bucket=bucket_name)
     elif backend == 'oci':
         return OCIObjectStore(bucket=bucket_name)
+    elif backend == 'uc':
```
Comment (karan6181) [inline composer/utils/file_helpers.py:352]: Is `uc` prefix a right way to go to check the the UC Volume path ? From what I know there are no such paths. Happy to sync up offline.

**Review by karan6181 — CHANGES_REQUESTED** [review]

File: composer/utils/file_helpers.py (hunk 2 above)
Comment (panchalhp-db) [reply composer/utils/file_helpers.py:352]: @karan6181 This is similar to what the streaming API uses: https://github.com/mosaicml/streaming/pull/361/files#diff-e29038e9a18c1ccfd11643e102e750e05acbb1eeb84f60b4198ab27ef705c7e0R422 we expect the user to specific a uc path as `uc://Volumes/catalog/schema/volume/`

File: composer/utils/object_store/uc_object_store.py (hunk 1 above)
Comment (panchalhp-db) [reply composer/utils/object_store/uc_object_store.py:15]: The current sdk that we are using does not have a list_objects API. Let me try to find out if there is a way to fetch this data. Were do we use list objects in the logging workflow?

File: composer/utils/object_store/uc_object_store.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        if not 'DATABRICKS_HOST' in os.environ or 'DATABRICKS_TOKEN' not in os.environ:
+            # TODO: Raise a better exception here
+            raise ValueError('Environment variables `DATABRICKS_HOST` and `DATABRICKS_TOKEN` '
+                             'must be set to use Databricks Unity Catalog Volumes')
+
+        from databricks.sdk import WorkspaceClient
+        self.client = WorkspaceClient()
+
+    @staticmethod
+    def _get_prefix(uri: str) -> str:
```
[the remark this answers is not in the document]
Comment (panchalhp-db) [reply composer/utils/object_store/uc_object_store.py:36]: Do you mean this specific function or just removing the prefix in general? The 'uc://` prefix that we have is mostly syntactic sugar introduced by us (similar to the one used in the streaming API) The path that the workspace clients understands is `/Volumes/<catalog>/<schema>/<volume>/path`

File: composer/utils/object_store/uc_object_store.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        # removeprefix works only with python3.9 and above
+        if hasattr(uri, 'removeprefix'):
+            return uri.removeprefix('uc:/')
+        else:
+            if uri.startswith('uc:/'):
+                return uri[len('uc:/'):]
+            return uri
+
+    def get_uri(self, object_name: str) -> str:
+        return f'uc:/{self.get_object_path(object_name)}'
```
[the remark this answers is not in the document]
Comment (panchalhp-db) [reply composer/utils/object_store/uc_object_store.py:46]: This is because `get_object_path(..)` returns something like `/Volumes/...` so `get_uri(..)` will return `uc://Volumes/...`

File: composer/utils/object_store/uc_object_store.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            except OSError:
+                pass
+            raise
+        else:
+            if overwrite:
+                os.replace(tmp_path, filename)
+            else:
+                os.rename(tmp_path, filename)
+
+    def get_object_size(self, object_name: str) -> int:
```
[the remark this answers is not in the document]
Comment (panchalhp-db) [reply composer/utils/object_store/uc_object_store.py:94]: Yeah I played around with the API a bit and found a way to do this. This should reliably work.

File: composer/utils/file_helpers.py (hunk 2 above)
Comment (panchalhp-db) [reply composer/utils/file_helpers.py:352]: Sync'ed offline with Karan. I'll change this to the latest recommendation from the UC Volumes team.

File: composer/utils/object_store/uc_object_store.py (hunk 1 above)
Comment (dakinggg) [reply composer/utils/object_store/uc_object_store.py:15]: we need to download all files from a folder (e.g. for the continued pretraining api)

File: tests/utils/object_store/test_uc_object_store.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        with open(file_to_download, 'wb') as fp:
+            fp.write(bytes('1' * (1024 * 1024 * 1024), 'utf-8'))
+        ws_client.files.download.side_effect = generate_dummy_file
+        uc_object_store.download_object(object_name, file_to_download, overwrite=True)
+        ws_client.files.download.assert_called_with('/Volumes/test-volume/remote-model.bin')
+
+        # verify that the file was actually overwritten
+        with open(file_to_download, 'rb') as f:
+            actual_content = f.readline()
+        assert actual_content == file_content
```
[the remark this answers is not in the document]
Comment (panchalhp-db) [reply tests/utils/object_store/test_uc_object_store.py:87]: It should be the same as the one on L56 right? We use the same variable to write the file in the callback on L59 too.

File: composer/utils/object_store/uc_object_store.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (panchalhp-db) [reply composer/utils/object_store/uc_object_store.py:36]: Changing this based on discussion with Karan

File: composer/utils/object_store/uc_object_store.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (panchalhp-db) [reply composer/utils/object_store/uc_object_store.py:46]: After the update, this will be of the format `dbfs:/Volumes/...` which is the expected format for dbfs files. I believe it is `dbfs:/` instead of `dbfs://` because it used to follow hadoop file format conventions although the latest hadoop url formats do except ``://` from what I can tell

[the remark this answers is not in the document]
Comment (dakinggg) [reply composer/utils/object_store/uc_object_store.py:46]: ok i think this works...
```
In [3]: parse_uri('dbfs:/Volumes/...')
Out[3]: ('dbfs', '', 'Volumes/...')
```
And I guess this makes sense because otherwise we think there is a bucket
```
In [4]: parse_uri('dbfs://Volumes/...')
Out[4]: ('dbfs', 'Volumes', '...')
```

**Comment by panchalhp-db** [comment]
@dakinggg @karan6181 I addressed your comments. This is ready for another pass. The existing sdk does not have support for listing files. I am waiting to hear back from the deco team that owns the SDK to see if there is an easy workaround to address this. I will follow up on this once I hear back. The rest of the code should be ready for another pass. Thanks!

File: composer/utils/file_helpers.py [hunk 7]
```diff
@@ -349,6 +349,8 @@ def maybe_create_object_store_from_uri(uri: str) -> Optional[ObjectStore]:
         return GCSObjectStore(bucket=bucket_name)
     elif backend == 'oci':
         return OCIObjectStore(bucket=bucket_name)
+    elif backend == 'dbfs' and path.startswith('Volumes/'):
+        return UCObjectStore(uri=uri)
     else:
         raise NotImplementedError(f'There is no implementation for the cloud backend {backend} via URI. Please use '
                                   's3 or one of the supported object stores')
```
Comment (karan6181) [inline composer/utils/file_helpers.py:356]: Not part of this PR, but we should change the exception message here. We can omit keyword `s3` since we support multiple cloud backends.

File: composer/utils/object_store/uc_object_store.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
+from __future__ import annotations
```
Comment (karan6181) [inline composer/utils/object_store/uc_object_store.py:1]: License header is missing. `pre-commit` adds it for you if you have missed it. So please run `pre-commit run -a` command before `git commit`. Thank you!

File: composer/utils/object_store/uc_object_store.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+    Args:
+        uri (str): The Databricks UC Volume URI that is of the format
+            `dbfs:/Volumes/<catalog-name>/<schema-name>/<volume-name>/path`
+
+    """
+
+    def __init__(self, uri: str) -> None:
+        try:
+            import databricks
```
Comment (karan6181) [inline composer/utils/object_store/uc_object_store.py:26]: Can we just replace `import databricks` with `from databricks.sdk import WorkspaceClient` since I don't see a usage of `databricks` directly ?

File: composer/utils/object_store/uc_object_store.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+        if not uri.startswith('dbfs:/Volumes'):
+            raise ValueError('Databricks Unity Catalog Volumes paths should start with "dbfs:/Volumes".')
+        self.path = uri.lstrip('dbfs:')
+
+        from databricks.sdk import WorkspaceClient
+        self.client = WorkspaceClient()
+
+    def get_uri(self, object_name: str) -> str:
+        return f'dbfs:{self.get_object_path(object_name)}'
```
Comment (karan6181) [inline composer/utils/object_store/uc_object_store.py:42]: nit missing method comments. Please also check other methods too! Thanks!

File: composer/utils/object_store/uc_object_store.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        if dirname:
+            os.makedirs(dirname, exist_ok=True)
+        tmp_path = str(filename) + f'{uuid.uuid4()}.tmp'
+
+        try:
+            from databricks.sdk.core import DatabricksError
+            try:
+                resp = self.client.files.download(self.get_object_path(object_name))
+                with open(tmp_path, 'wb') as f:
+                    f.write(resp.contents.read())
```
Comment (karan6181) [inline composer/utils/object_store/uc_object_store.py:78]: Depending on how big is the file, if you read all the content at once, it will load into RAM and there is not enough RAM, then it will fail with OOM issue. I would encourage to read chunk by chunk, the chunk size can set to any value, such as in 4MB, 64MB, 512MB, etc. Downside of this approach is that it would be little slower than reading at once.

File: composer/utils/file_helpers.py [hunk 12]
```diff
@@ -349,9 +349,11 @@ def maybe_create_object_store_from_uri(uri: str) -> Optional[ObjectStore]:
         return GCSObjectStore(bucket=bucket_name)
     elif backend == 'oci':
         return OCIObjectStore(bucket=bucket_name)
+    elif backend == 'dbfs' and path.startswith('Volumes/'):
```
Comment (dakinggg) [inline composer/utils/file_helpers.py:352]: same, separate error here please

File: composer/utils/file_helpers.py [hunk 13]
```diff
@@ -387,10 +389,11 @@ def maybe_create_remote_uploader_downloader_from_uri(
     elif backend == 'wandb':
         raise NotImplementedError(f'There is no implementation for WandB via URI. Please use '
                                   'WandBLogger with log_artifacts set to True')
-
+    elif backend == 'dbfs' and path.startswith('Volumes/'):
+        return RemoteUploaderDownloader(bucket_uri=f'{backend}://{bucket_name}', backend_kwargs={uri: uri})
```
Comment (dakinggg) [inline composer/utils/file_helpers.py:393]: bucket name here is empty string right? should we just leave it out of the format string for clarity or is there a reason it is there?

File: composer/utils/object_store/uc_object_store.py [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+        Args:
+            object_name (str): The name of the object to download i.e. path relative to the root of the volume.
+            filename (str | pathlib.Path): The local path where a the file needs to be downloaded.
+            overwrite(bool, optional): Whether to overwrite an existing file at ``filename``, if it exists.
+                (default: ``False``)
+            callback ((int) -> None, optional): Unused
+
+        Raises:
+            FileNotFoundError: If the file was not found in UC volumes.
+            DatabricksError: If there was any other error querying the Databricks UC volumes.
```
Comment (dakinggg) [inline composer/utils/object_store/uc_object_store.py:112]: do we want to retry these errors? If so, we need to reraise as `ObjectStoreTransientError`

File: composer/utils/object_store/uc_object_store.py [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+        try:
+            import databricks
+        except ImportError as e:
+            raise MissingConditionalImportError('databricks') from e
+
+        if not 'DATABRICKS_HOST' in os.environ or not 'DATABRICKS_TOKEN' in os.environ:
+            raise ValueError('Environment variables `DATABRICKS_HOST` and `DATABRICKS_TOKEN` '
+                             'must be set to use Databricks Unity Catalog Volumes')
+
+        if not uri.startswith('dbfs:/Volumes'):
```
[the remark this answers is not in the document]
Comment (panchalhp-db) [reply composer/utils/object_store/uc_object_store.py:54]: Okay makes sense. let me fix this!

File: tests/utils/object_store/test_uc_object_store.py [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+
+
+def test_uc_object_store_without_env():
+    with pytest.raises(ValueError):
+        UCObjectStore(uri='dbfs:/Volumes/test-volume/')
+
+
+def test_uc_object_store_invalid_uri():
+    with pytest.raises(ValueError):
+        UCObjectStore(uri='dbfs:/root/')
```
[the remark this answers is not in the document]
Comment (panchalhp-db) [reply tests/utils/object_store/test_uc_object_store.py:38]: Ah good catch. Let me fix.

**Review by dakinggg — COMMENTED** [review]
This looks good to me! Can you please create a jira for adding log_model support? You can put it in the genai runtime project and the mosaic finetuning private preview epic.

and @eracah can you please do a pass and approve?

**Comment by panchalhp-db** [comment]
> This looks good to me! Can you please create a jira for adding log_model support? You can put it in the genai runtime project and the mosaic finetuning private preview epic.
> 
> and @eracah can you please do a pass and approve?

Thank you @dakinggg I added jira tickets for:
- log_model support: https://databricks.atlassian.net/browse/MCLOUD-3130
- list_objects support: https://databricks.atlassian.net/browse/MCLOUD-3130

**Review by eracah — COMMENTED** [review]
Looks pretty good. Can you add a small e2e test, where you actually train for a few batches and save to UC then load from it. It can be a manual test. [Here's](https://github.com/mosaicml/composer/blob/dev/tests/utils/object_store/test_gs_object_store.py#L28-L51) an example

**Comment by panchalhp-db** [comment]
> Looks pretty good. Can you add a small e2e test, where you actually train for a few batches and save to UC then load from it. It can be a manual test. [Here's](https://github.com/mosaicml/composer/blob/dev/tests/utils/object_store/test_gs_object_store.py#L28-L51) an example

Thank you @eracah I updated the PR with the e2e test and also fixed some bugs that I found on the way. UC Volume treats the first 4 directories of the path as the bucket root so we have to handle cases where `object_name` is passed relative to bucket root eg: `.credentials_validated_successfully` as well as cases where the full path like `Volumes/ml/mosaicml/test-volume/checkpoints/1695720685-tourmaline-wolf/test-model.pt` is passed. I added tests and comments to document this behavior. PTAL. cc: @dakinggg  too

File: composer/utils/file_helpers.py [hunk 17]
```diff
@@ -349,9 +349,15 @@ def maybe_create_object_store_from_uri(uri: str) -> Optional[ObjectStore]:
         return GCSObjectStore(bucket=bucket_name)
     elif backend == 'oci':
         return OCIObjectStore(bucket=bucket_name)
+    elif backend == 'dbfs':
+        if path.startswith('Volumes/'):
```
Comment (dakinggg) [inline composer/utils/file_helpers.py:353]: I think the helpers here should also validate the 4 dirs thing? maybe `UCObjectStore` should expose a static method that just validates a path is the right format instead of duplicating here

**Review by dakinggg — APPROVED** [review]
LGTM, thanks Harsh!

File: tests/utils/object_store/test_uc_object_store.py [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+def uc_object_store(ws_client, monkeypatch):
+    db = pytest.importorskip('databricks.sdk', reason='requires databricks')
+
+    monkeypatch.setenv('DATABRICKS_HOST', 'test-host')
+    monkeypatch.setenv('DATABRICKS_TOKEN', 'test-token')
+    with mock.patch.object(db, 'WorkspaceClient', lambda: ws_client):
+        yield UCObjectStore(path='Volumes/catalog/schema/volume/path/')
+
+
+@pytest.mark.remote  # databricks auth is hard to set up on github actions
```
Comment (dakinggg) [inline tests/utils/object_store/test_uc_object_store.py:36]: this won't pass. "remote" means "run only on github actions, not on local machine". Could you make a jira for getting databricks auth into github actions? you can just skip this test for now, and we will rely on the manual test of it that you did

**Review by eracah — APPROVED** [review]
LGTM since you ran the e2e test manually then I'm satisfied!

File: tests/utils/object_store/test_uc_object_store.py (hunk 18 above)
Comment (panchalhp-db) [reply tests/utils/object_store/test_uc_object_store.py]: Ah I see. Added a ticket here: https://databricks.atlassian.net/browse/MCLOUD-3169 and skipped the test for now.

**Comment by mvpatel2000** [comment]
This PR is adding _really_ slow tests to CPU. Can we slim them down before merging?
```
============================= slowest 20 durations =============================
21.57s call     tests/utils/object_store/test_uc_object_store.py::test_download_object[overwrite_file]
14.15s call     tests/loggers/test_mlflow_logger.py::test_mlflow_log_model
11.10s call     tests/utils/object_store/test_uc_object_store.py::test_download_object[success]
7.72s call     composer/algorithms/factorize/README.md::line 57
7.65s call     composer/algorithms/ghost_batchnorm/README.md::line 52
7.64s call     docs/source/method_cards/ghost_batchnorm.md::line 52
7.62s call     docs/source/method_cards/factorize.md::line 57
7.25s call     README.md::line 108
6.74s call     tests/datasets/test_mnist.py::test_mnist_shape_length[True-True]
6.68s call     tests/utils/object_store/test_uc_object_store.py::test_download_object[file_exists]
6.52s call     tests/trainer/test_checkpoint.py::TestCheckpointSaving::test_s3_uri_creates_remote_ud[True]
6.43s call     tests/trainer/test_trainer_eval.py::test_eval_dur_interval_token_max[1-1-0.57dur-512tok]
```

**Comment by panchalhp-db** [comment]
> This PR is adding _really_ slow tests to CPU. Can we slim them down before merging?
> 
> ```
> ============================= slowest 20 durations =============================
> 21.57s call     tests/utils/object_store/test_uc_object_store.py::test_download_object[overwrite_file]
> 14.15s call     tests/loggers/test_mlflow_logger.py::test_mlflow_log_model
> 11.10s call     tests/utils/object_store/test_uc_object_store.py::test_download_object[success]
> 7.72s call     composer/algorithms/factorize/README.md::line 57
> 7.65s call     composer/algorithms/ghost_batchnorm/README.md::line 52
> 7.64s call     docs/source/method_cards/ghost_batchnorm.md::line 52
> 7.62s call     docs/source/method_cards/factorize.md::line 57
> 7.25s call     README.md::line 108
> 6.74s call     tests/datasets/test_mnist.py::test_mnist_shape_length[True-True]
> 6.68s call     tests/utils/object_store/test_uc_object_store.py::test_download_object[file_exists]
> 6.52s call     tests/trainer/test_checkpoint.py::TestCheckpointSaving::test_s3_uri_creates_remote_ud[True]
> 6.43s call     tests/trainer/test_trainer_eval.py::test_eval_dur_interval_token_max[1-1-0.57dur-512tok]
> ```

@mvpatel2000 great catch.. I was downloading a 1Gb file in my tests 🤦🏾‍♂️ which is why this was slow. Fixing it now. Thank you!
